### PR TITLE
refactor(api): allow load_module and load_labware via PE commands

### PR DESCRIPTION
**Pull request by TamarZanzouri**
# Overview

work related to https://opentrons.atlassian.net/browse/RCORE-257.
allow load_module and load_labware via module_core PE commands.

# Changelog

- moved `protocol_core.get_deck()` and `geometry.add_labware` to add_labware_core LegacyModuleCore
- changed add_labware_core to no-op
- implemented module_core.get_deck_slot()
- added abstract implementations for all engine modules. 
- implemented engine protocol.load_module()
- implemented engine protocol.load_labware() with a module as location

# Review requests

1. is the implementation correct?
2. did I miss anything?
3. am I missing tests for this? 

tested with dev server all modules.

### CODE REVIEW

File: api/src/opentrons/protocol_api/core/engine/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+from typing import Optional
+
+
+class InvalidModuleLocationError(ValueError):
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_api/core/engine/exceptions.py:6]: Should I move this somewhere else?

File: api/src/opentrons/protocol_api/protocol_context.py [hunk 2]
```diff
@@ -680,14 +685,18 @@ def _create_module_context(
...
 ) -> ModuleTypes:
-    # TODO(mc, 2022-09-07): create distinct module cores
-    module_constructors: Dict[ModuleType, Type[ModuleTypes]] = {
-        ModuleType.MAGNETIC: MagneticModuleContext,
-        ModuleType.TEMPERATURE: TemperatureModuleContext,
-        ModuleType.THERMOCYCLER: ThermocyclerContext,
-        ModuleType.HEATER_SHAKER: HeaterShakerContext,
-    }
-    module_cls = module_constructors[module_core.get_type()]
+    print(module_core)
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_api/protocol_context.py]:
```suggestion
```

File: api/src/opentrons/protocol_api/protocol_context.py [hunk 3]
```diff
@@ -680,14 +685,17 @@ def _create_module_context(
...
+    module_cls: Optional[Type[ModuleTypes]] = None
+    if isinstance(module_core, AbstractTemperatureModuleCore):
+        module_cls = TemperatureModuleContext
+    elif isinstance(module_core, AbstractMagneticModuleCore):
+        module_cls = MagneticModuleContext
+    elif isinstance(module_core, AbstractThermocyclerCore):
+        module_cls = ThermocyclerContext
+    elif isinstance(module_core, AbstractHeaterShakerCore):
+        module_cls = HeaterShakerContext
+    else:
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_api/protocol_context.py:697]: Should I remove this else?

File: api/src/opentrons/protocol_api/core/engine/protocol.py [hunk 4]
```diff
@@ -108,11 +134,39 @@ def load_labware(
...
         self,
         model: ModuleModel,
-        location: Optional[DeckSlotName],
+        deck_slot: Optional[DeckSlotName],
         configuration: Optional[str],
     ) -> ModuleCore:
         """Load a module into the protocol."""
-        raise NotImplementedError("ProtocolEngine PAPI core not implemented")
+        if deck_slot is None:
+            if ModuleType.from_model(model) == ModuleType.THERMOCYCLER:
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/engine/protocol.py:144]: Can we rename this to `from_hardware_model`?

File: api/src/opentrons/protocol_api/core/engine/protocol.py [hunk 5]
```diff
@@ -108,11 +134,39 @@ def load_labware(
     def load_module(
         self,
         model: ModuleModel,
-        location: Optional[DeckSlotName],
+        deck_slot: Optional[DeckSlotName],
         configuration: Optional[str],
     ) -> ModuleCore:
         """Load a module into the protocol."""
-        raise NotImplementedError("ProtocolEngine PAPI core not implemented")
+        if deck_slot is None:
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/engine/protocol.py:143]:
```suggestion
        # TODO(mc, 2022-10-20): move to public ProtocolContext
        # once `Deck` and `ProtocolEngine` play nicely together
        if deck_slot is None:
```

File: api/src/opentrons/protocol_api/core/engine/protocol.py [hunk 6]
```diff
@@ -108,11 +134,39 @@ def load_labware(
...
+        if module_type == ModuleType.TEMPERATURE:
+            module_core_cls = TemperatureModuleCore
+        elif module_type == ModuleType.MAGNETIC:
+            module_core_cls = MagneticModuleCore
+        elif module_type == ModuleType.THERMOCYCLER:
+            module_core_cls = ThermocyclerModuleCore
+        elif module_type == ModuleType.HEATER_SHAKER:
+            module_core_cls = HeaterShakerModuleCore
+        else:
+            assert False, "Unsupported module definition"
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/engine/protocol.py]: Let's remove this

File: api/src/opentrons/protocol_api/core/engine/module_core.py [hunk 7]
```diff
@@ -1,7 +1,326 @@
...
+                                  command after ``temperature`` is reached.
+        :param block_max_volume: The maximum volume of any individual well
+                                 of the loaded labware. If not supplied,
+                                 the thermocycler will default to 25µL/well.
+
+        .. note:
+
+            If ``hold_time_seconds`` is not specified, the Thermocycler
+            will proceed to the next command after ``temperature`` is reached.
+        """
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/engine/module_core.py]: Same comment as #11550: should be Google-style docstring

File: api/src/opentrons/protocol_api/core/engine/module_core.py [hunk 8]
```diff
@@ -1,7 +1,326 @@
...
+                                 of the loaded labware. If not supplied,
+                                 the thermocycler will default to 25µL/well.
+
+        .. note:
+
+            Unlike the :py:meth:`set_block_temperature`, either or both of
+            'hold_time_minutes' and 'hold_time_seconds' must be defined
+            and finite for each step.
+
+        """
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/engine/module_core.py]: Same comment as #11550: should be Google-style docstring

File: api/src/opentrons/protocol_api/core/engine/exceptions.py (hunk 1 above)
Comment (mcous) [inline api/src/opentrons/protocol_api/core/engine/exceptions.py:6]: Add same TODO about how this logic needs to move to the public ProtocolContext

File: api/src/opentrons/protocol_api/core/module.py [hunk 9]
```diff
@@ -52,7 +53,7 @@ def get_deck_slot(self) -> DeckSlotName:
         """Get the module's deck slot."""
 
     @abstractmethod
-    def add_labware_core(self, labware_core: LabwareCoreType) -> None:
+    def add_labware_core(self, labware_core: LabwareCoreType) -> Labware:
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/module.py:57]:
```suggestion
    # TODO(mc, 2022-10-20): make this a side-effect, not a return
    def add_labware_core(self, labware_core: LabwareCoreType) -> Labware:
```

File: api/src/opentrons/protocol_api/protocol_context.py (hunk 3 above)
Comment (mcous) [reply api/src/opentrons/protocol_api/protocol_context.py:697]: No need

File: api/tests/opentrons/protocol_api/core/engine/test_module_core.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
+"""Tests for opentrons.protocol_api.core.engine.ModuleCore."""
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/engine/test_module_core.py:1]: Add tests for other methods we implemented

File: api/tests/opentrons/protocol_api/core/engine/test_protocol_core.py [hunk 11]
```diff
@@ -27,7 +55,7 @@ def mock_engine_client(decoy: Decoy) -> EngineClient:
 @pytest.fixture
 def subject(mock_engine_client: EngineClient) -> ProtocolCore:
     """Get a ProtocolCore test subject with its dependencies mocked out."""
-    return ProtocolCore(engine_client=mock_engine_client)
+    return ProtocolCore(engine_client=mock_engine_client, api_version=APIVersion(2, 12))
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/engine/test_protocol_core.py]: Add test for `api_version` property

File: api/src/opentrons/protocol_api/core/engine/module_core.py [hunk 12]
```diff
@@ -1,7 +1,326 @@
...
+
+    def __init__(
+        self,
+        module_id: str,
+        engine_client: ProtocolEngineClient,
+        api_version: APIVersion,
+    ) -> None:
+        self._module_id = module_id
+        self._engine_client = engine_client
+        # tz,(10-20-22): if the only method that uses this is the add_labware_core figure out a better way to pass this along.
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/engine/module_core.py]: Can remove note, I think having it in `init` is correct

**Review by mcous — COMMENTED** [review]
Minor comments, looks fundamentally good

File: api/src/opentrons/protocol_api/core/engine/protocol.py (hunk 4 above)
Comment (TamarZanzouri) [reply api/src/opentrons/protocol_api/core/engine/protocol.py:144]: done! leaving this for transparency

**Comment by TamarZanzouri** [comment]
Tested on at OT-2, all modules work! One note we still need to implement but not for this PR is disallowing to upload a thermocycler on a slot that is not 7.

File: api/src/opentrons/hardware_control/api.py [hunk 13]
```diff
@@ -1071,7 +1071,7 @@ async def create_simulating_module(
         return await self._backend.module_controls.build_module(
             port="",
             usb_port=USBPort(name="", port_number=0),
-            type=modules.ModuleType.from_model(model),
+            type=modules.ModuleType.from_hardware_model(model),
```
Comment (mcous) [inline api/src/opentrons/hardware_control/api.py]: Oh shoot, I got confused in my earlier comment. I thought that this was the `protocol_engine`'s `ModuleType` enum. There's no need for the `hardware_control`'s `ModuleType` enum to be named like this; we should revert this change

File: api/src/opentrons/protocol_api/core/engine/module_core.py [hunk 14]
```diff
@@ -1,7 +1,355 @@
...
+        """Add a labware to the module."""
+        return Labware(implementation=labware_core, api_level=self._api_version)
+
+
+class TemperatureModuleCore(ModuleCore, AbstractTemperatureModuleCore[LabwareCore]):
+    """Temperature Module core logic implementation for Python protocols."""
+
+    def set_target_temperature(self, celsius: float) -> None:
+        """Set the Temperature Module's target temperature in °C."""
+        raise NotImplementedError("set_target_temperature not implemented")
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/engine/module_core.py:97]: Follow this pattern for all other cores

File: api/src/opentrons/protocol_api/core/engine/protocol.py [hunk 15]
```diff
@@ -108,11 +134,39 @@ def load_labware(
...
+
+        module_core_cls: Type[ModuleCore] = ModuleCore
+        if module_type == ModuleType.TEMPERATURE:
+            module_core_cls = TemperatureModuleCore
+        elif module_type == ModuleType.MAGNETIC:
+            module_core_cls = MagneticModuleCore
+        elif module_type == ModuleType.THERMOCYCLER:
+            module_core_cls = ThermocyclerModuleCore
+        elif module_type == ModuleType.HEATER_SHAKER:
+            module_core_cls = HeaterShakerModuleCore
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/engine/protocol.py]: Add TODO: move to module core factory function

File: api/src/opentrons/protocol_api/core/protocol_api/legacy_module_core.py [hunk 16]
```diff
@@ -95,9 +95,11 @@ def get_deck_slot(self) -> DeckSlotName:
         """Get the module's deck slot."""
         return DeckSlotName.from_primitive(self._geometry.parent)  # type: ignore[arg-type]
 
-    def add_labware_core(self, labware_core: LabwareImplementation) -> None:
+    def add_labware_core(self, labware_core: LabwareImplementation) -> Labware:
         """Add a labware to the module."""
-        pass
+        labware = self.geometry.add_labware(Labware(implementation=labware_core))
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/protocol_api/legacy_module_core.py:101]: Add TODO link to RSS-105 and RSS-106, which forces us to return a `Labware` from the method, which we don't want to do

File: api/tests/opentrons/protocol_api/core/engine/test_module_core.py [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+from opentrons.types import DeckSlotName
+
+
+@pytest.fixture
+def mock_engine_client(decoy: Decoy) -> EngineClient:
+    """Get a mock ProtocolEngine synchronous client."""
+    return decoy.mock(cls=EngineClient)
+
+
+@pytest.fixture()
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/engine/test_module_core.py]:
```suggestion
@pytest.fixture
```

File: api/tests/opentrons/protocol_api/core/engine/test_module_core.py [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+def subject(mock_engine_client: EngineClient) -> ModuleCore:
+    """Get a ModuleCore test subject."""
+    return ModuleCore(
+        module_id="1234",
+        engine_client=mock_engine_client,
+        api_version=APIVersion(2, 13),
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/engine/test_module_core.py]:
```py
from opentrons.protocol_api import MAX_SUPPORTED_VERSION
...
api_version=MAX_SUPPORTED_VERSION,
```

File: api/tests/opentrons/protocol_api/core/engine/test_module_core.py [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+    )
+
+
+def test_get_deck_slot(
+    decoy: Decoy, subject: ModuleCore, mock_engine_client: EngineClient
+) -> None:
+    """Should return the deck slot accosiated to the module id."""
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/engine/test_module_core.py]:
```suggestion
    """Should return the deck slot associated to the module id."""
```

File: api/tests/opentrons/protocol_api/core/engine/test_module_core.py [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+    decoy.when(mock_engine_client.state.modules.get_location("1234")).then_return(
+        DeckSlotLocation(slotName=DeckSlotName.SLOT_1)
+    )
+
+    assert subject.get_deck_slot() == DeckSlotName.SLOT_1
+
+
+def test_add_labware_core(decoy: Decoy, subject: ModuleCore) -> None:
+    """Should return a Labware obejct."""
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/engine/test_module_core.py]:
```suggestion
    """Should return a Labware object."""
```

File: api/tests/opentrons/protocol_api/core/engine/test_module_core.py [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+) -> None:
+    """Should return the deck slot accosiated to the module id."""
+    decoy.when(mock_engine_client.state.modules.get_location("1234")).then_return(
+        DeckSlotLocation(slotName=DeckSlotName.SLOT_1)
+    )
+
+    assert subject.get_deck_slot() == DeckSlotName.SLOT_1
+
+
+def test_add_labware_core(decoy: Decoy, subject: ModuleCore) -> None:
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/engine/test_module_core.py]: TODO (can be comment): add `api_version` fixture

File: api/tests/opentrons/protocol_api/core/engine/test_module_core.py [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+@pytest.fixture()
+def subject(mock_engine_client: EngineClient) -> ModuleCore:
+    """Get a ModuleCore test subject."""
+    return ModuleCore(
+        module_id="1234",
+        engine_client=mock_engine_client,
+        api_version=APIVersion(2, 13),
+    )
+
+
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/engine/test_module_core.py:37]: Missing test for `api_version` property

File: api/tests/opentrons/protocol_api/core/engine/test_protocol_core.py [hunk 23]
```diff
@@ -100,3 +170,146 @@ def test_add_labware_definition(
     result = subject.add_labware_definition(minimal_labware_def)
 
     assert result == LabwareLoadParams("hello", "world", 123)
+
+
+@pytest.mark.parametrize(
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/engine/test_protocol_core.py:188]: Same TODO as "create a ModuleCore factory function"

**Review by mcous — APPROVED** [review]
Minor comments, can be addressed in a follow up PR or in this one; whatever's most convenient for speed

**Review by jbleon95 — APPROVED** [review]
